packages/polling-controller/src/BlockTrackerPollingController.ts

```diff
@@ -0,0 +1,88 @@
+import { BaseController, BaseControllerV1 } from '@metamask/base-controller';
+import type {
+  NetworkClientId,
+  NetworkClient,
+} from '@metamask/network-controller';
+import type { Json } from '@metamask/utils';
+
+import {
+  AbstractPollingControllerBaseMixin,
+  getKey,
+} from './AbstractPollingController';
+import type { PollingTokenSetId } from './AbstractPollingController';
+
+type Constructor = new (...args: any[]) => object;
+
+/**
+ * BlockTrackerPollingControllerMixin
+ * A polling controller that polls using a block tracker.
+ *
+ * @param Base - The base class to mix onto.
+ * @returns The composed class.
+ */
+function BlockTrackerPollingControllerMixin<TBase extends Constructor>(
+  Base: TBase,
+) {
+  abstract class BlockTrackerPollingController extends AbstractPollingControllerBaseMixin(
+    Base,
+  ) {
+    #activeListeners: Record<string, (options: Json) => Promise<void>> = {};
+
+    abstract _getNetworkClientById(
+      networkClientId: NetworkClientId,
+    ): NetworkClient | undefined;
+
+    _startPollingByNetworkClientId(
+      networkClientId: NetworkClientId,
+      options: Json,
+    ) {
+      const key = getKey(networkClientId, options);
+
+      if (this.#activeListeners[key]) {
+        return;
+      }
+
+      const networkClient = this._getNetworkClientById(networkClientId);
+      if (networkClient && networkClient.blockTracker) {
+        console.log('networkClient', networkClient);
+        const updateOnNewBlock = this._executePoll.bind(
+          this,
+          networkClientId,
+          options,
+        );
+        networkClient.blockTracker.addListener('latest', updateOnNewBlock);
+        this.#activeListeners[key] = updateOnNewBlock;
+      } else {
+        throw new Error(
+          `Unable to retrieve blockTracker for networkClientId ${networkClientId}`,
+        );
+      }
+    }
+
+    _stopPollingByPollingTokenSetId(key: PollingTokenSetId) {
+      const [networkClientId] = key.split(':');
+      const networkClient = this._getNetworkClientById(
+        networkClientId as NetworkClientId,
+      );
+
+      if (networkClient && this.#activeListeners[key]) {
```

Comment:
should we check for existence of blockTracker here if we also checked it in `_startPollingByNetworkClientId`? Do we need to check it at all?

Reply:
We don't need to check this in `_startPollingByNetworkClientId` since its a required field for the `NetworkClient`. Removed that check here https://github.com/MetaMask/core/pull/3636/commits/09e9e30ceea6d1f283899f016e40c97e2de90cb4